Add git-blame-historic-context skill (#4373)
cowprotocol/services: docs/skills/git-blame-historic-context.md

# Skill — `git blame` for historic context

> Use before flagging code that looks unusual, redundant, accidental, or "easy to clean up". Often that code looks weird because it had to.

Reviewer: This or similar
```suggestion
Use before flagging code that looks unusual, redundant, accidental, or "easy to clean up". Often times, certain decisions led to sub-optimal code, those decisions are also often codified in git history.
```

## How to invoke

Two ways:

> - **Procedurally** — follow the steps below when you encounter unusual-looking code in any context (PR review, order-debug session, ad-hoc *"why is this here?"*).

Reviewer: Not sure how this interacts with other parts, but this shouldn't be invoked the first time the agent comes around weird code. 

The agent needs to have a full picture of the PR they're analyzing first, gather places where code "looks off" while it analyzed it and only then, run this. Often times, the code that looks off is being partially fix, moved, etc

Author: added a 'when to invoke' section in the PR, for when it's being used inside the pr-review scenario.
also, for context: this rule was in the main /review-pr skill, where it waited for full context, before invoking this git blame check; when something in the existing code, didn't make complete sense

- **Via slash command** — `/blame-context <path>:<line>` wraps the procedure end-to-end and prints the strengthens / weakens / inverts decision. Useful when you want a one-shot answer rather than walking the procedure manually.

## Example

A reviewer sees this in `crates/driver/src/domain/competition/solution/settlement.rs:444`:

```rust
let max_gas = eth::Gas(block_limit.0 / eth::U256::from(2));
```

`/2` looks arbitrary — a reviewer might be tempted to flag it as a magic number. Run the procedure first:

```bash
git blame -L 444,444 -- crates/driver/src/domain/competition/solution/settlement.rs
# → a4ee76aae3  Felix Leupold  2024-03-18  ...

git log -1 --format='%s%n%b' a4ee76aae3
# → subject ends with `(#NNNN)`; pivot to the PR
gh pr view <NNNN>
# → body: "block builders' default algorithm picks tx whose gas limit
#    fits remaining space; leave headroom for inclusion."
```

> Decision: **weakens** the "magic number" suspicion — the constant has documented inclusion-economics rationale. Drop the finding, or downgrade to a `Question:` confirming the rationale still applies on the chain in question.

Reviewer: I think you need to find a second example, not to replace, but rather to provide more than one data point 

I find that this example is a bit lackluster

## Procedure

```bash
git blame -L <start>,<end> -- <path>                # who/what/when

# cowprotocol/services squash-merges. Most blames point at one commit whose
# subject ends with "(#NNNN)" — extract the PR number, then pivot to the
# PR conversation, which is usually richer than the commit body alone.
git log -1 --format='%s%n%b' <sha>
gh pr view <NNNN>
```

## Decision

> Promote what blame reveals into the finding's Explanation, then weight the finding:

Reviewer: ```suggestion
Promote what blame reveals into the finding's Explanation, then weigh the finding:
```

- **Strengthens** — surrounding code was added recently for a reason the diff now contradicts. Keep / raise severity.
- **Weakens** — the originating PR explains *why* the shape is unusual (deliberate workaround, perf fix, cross-version compat). Soften, or pivot from `Action:` to `Question:`. Example: a two-line guard looked over-defensive; blame showed it was a hot-patch lifted from prod logs — Medium → Question, asking whether the failure mode still applies.
- **Inverts** — flagging this would ask the author to undo a hard-won fix. Drop the finding. Example: a `Some(_) =>` arm looked redundant; blame revealed it was added months ago to swallow a panic on an edge case the diff was about to remove. Dropped, with a note that the panic is back.

## Edge cases

> - **Merge commit, not squash** — rare here; inspect with `git log -1 --format='%s%n%b' <sha>` and walk parents (`<sha>^1`, `^2`) to find the commit that actually authored the line.

Reviewer: No need to pollute context with rarity comments 
```suggestion
- **Merge commit, not squash** — inspect with `git log -1 --format='%s%n%b' <sha>` and walk parents (`<sha>^1`, `^2`) to find the commit that actually authored the line.
```
